### PR TITLE
🔎 fix: Report Measured Skill Catalog Truncation

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

`injectSkillCatalog` warns when a skill's description will not reach the model verbatim. The warning assumed `SKILL_CATALOG_MAX_ENTRY_CHARS` (250) was the delivered length. It isn't — 250 is only the first rung of the SDK's truncation ladder.

`formatSkillCatalog` in `@librechat/agents` applies the per-entry cap, then truncates **further** proportionally against its own context budget (`contextWindowTokens × 0.01 × 4` chars), and finally falls back to names-only, dropping descriptions entirely.

That leaves two silent failure modes:

- **A description *under* the cap can still be cut, with no warning at all.** Measured against the real SDK: 8 skills × 200 chars at a 20k context window each reach the model at 85 chars. The old check (`length > 250`) never fires.
- **When the ladder engages past the cap, the warning overstated what landed.** A 499-char description in a 100-skill catalog was reported as "truncated to 250" when 60 chars actually reached the model.

Measured 2026-09-04 at a 200k window: ≤26 skills get the full 250, 50 skills get 141, 100 (the catalog limit) get 61. A smaller context window tightens it independently — the same six-skill catalog gets 189 chars each at 32k, 29 at 8k.

## Change

Measure delivered lengths on a **length-equivalent probe** rather than on the real catalog.

Every decision in `formatSkillCatalog`'s truncation ladder reads description `.length` and never description content (`skillCatalog.ts` lines 51-89). So formatting a second catalog whose descriptions are same-length filler reproduces the real cuts exactly, and that probe is parseable by construction: filler carries no newline and no entry marker, and skill names are validated to `^[a-z0-9][a-z0-9-]*$`.

This arrived by elimination. Four review rounds showed the real catalog cannot be parsed or matched, because descriptions are arbitrary text that can imitate any structure a check relies on:

- **Keying by name** collapsed duplicate-named skills, which the catalog deliberately keeps as separate entries — a 400-char and a 100-char description sharing a name warned "truncated to 100 of 400" when the first actually reached 250.
- **Parsing physical lines** broke on multiline descriptions, which nothing strips (`validateSkillDescription` checks type, length and trim-emptiness only). One kept intact was flagged truncated; one starting with a newline measured 0 and read as dropped; and a continuation line imitating the next entry's marker desynchronized everything after it, measuring a 400-char entry at 8.
- **Matching the description text** missed a drop whenever that text occurred elsewhere. Under names-only, a one-word description such as `research` collides with a skill *named* `research`, so the check held for a description that had been dropped entirely.
- **Matching the whole entry rendering** still missed the case where truncation splices one entry's tail onto the next: cut a description at `…\n- next-skill: next description` with a real matching `next-skill`, and the full rendering is present across two entries.

Measuring on the probe closes the class rather than the instance, and keeps the delivered length in the warning.

Model-facing behaviour is unchanged throughout: this is an observability fix. The catalog text, the cap, and the injected instructions are untouched.

## On the mock

`formatSkillCatalog` is no longer stubbed in the spec. The stub was a passthrough that never truncated, which is exactly why no existing test could observe this class of defect. The SDK exports the real function, so the spec now exercises it — per the repo's "real logic over mocks" guidance.

## Testing

`packages/api/src/agents/__tests__/skills.test.ts` — 163 passed. Coverage added for over-cap truncation, sub-cap budget truncation, the names-only fallback, duplicate names, multiline and leading-newline descriptions, the imitator case, the name-collision drop, and the spliced-boundary case.

Test selection came from codegraph (`tests.mjs`), which put this spec at depth 1 and everything else at depth ≥ 2.

The truncation tests were confirmed to **fail** with the fix reverted, so they pin the behaviour rather than passing either way. Every approach was also diffed against the real formatter across the six adversarial cases raised in review: probe measurement correct on 6/6, entry-scoped containment on 5/6, bare containment on 4/6, positional parsing on 3/6.

`npx tsc --noEmit` in `packages/api` reports 30 errors, identical to the `dev` baseline on a clean tree — all in `memory.*`, `management.*` and `telemetry/logs.*`, none in touched files.

## Not addressed here

Whether 250 is still the right per-entry number is a separate design question, left alone deliberately. Worth noting the SDK already implements a per-catalog token budget — the ladder — so LibreChat's fixed 250 is a per-entry ceiling feeding into it, not the budget itself.

Companion docs fix, which is where authors actually learn this: danny-avila/codegraph-poc#4

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 1]
```diff
@@ -583,6 +585,33 @@ export async function resolveSkillCatalog(
...
+    if (!line.startsWith('- ')) {
+      continue;
+    }
+    const separator = line.indexOf(': ');
+    if (separator === -1) {
+      continue;
+    }
+    const name = line.slice(2, separator);
+    if (names.has(name)) {
+      delivered.set(name, line.length - separator - 2);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Confirmed and fixed in 906462b — real defect, good catch.

Reproduced exactly as described: a 400-char and a 100-char description sharing one name warned `truncated to 100 of 400` when the first entry actually reached the model at 250.

Switched to positional matching. Entry order is the order passed to the formatter, and the names-only fallback drops from the tail, so a skill the catalog never reached keeps its `0` and reads as dropped. Regression test added — verified to fail against the name-keyed version with `Received: 100`, the value you predicted.

File: packages/api/src/agents/skills.ts [hunk 2]
```diff
@@ -583,6 +585,42 @@ export async function resolveSkillCatalog(
...
+ * deliberately keeps duplicate-named skills as separate entries; keying by
+ * name would collapse them and misreport every occurrence but the last.
+ * Entry order is the order passed to the formatter, and the names-only
+ * fallback drops from the tail, so a skill the catalog never reached keeps
+ * its `0` and reads as dropped.
+ */
+function measureCatalogDescriptions(catalog: string, names: string[]): number[] {
+  const delivered = new Array<number>(names.length).fill(0);
+  let index = 0;
+  for (const line of catalog.split('\n')) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Confirmed and fixed in 4e2223c. Verified all three modes against the real formatter:

- multiline kept intact -> falsely flagged truncated
- leading newline -> measured 0, falsely flagged dropped
- continuation imitating the next marker -> desync, a 400-char entry measured 8

You're right that entry boundaries aren't recoverable: a description may contain the exact byte sequence that starts the next entry, so no parse can disambiguate. Rather than normalize, I dropped the line parsing entirely — a description the catalog kept appears verbatim, so containment detects truncation from any rung of the ladder without needing boundaries, and still handles the duplicate-name case from your first finding.

Cost is the delivered-length figure in the warning. That number is what kept being wrong across both rounds, and the actionable signal is that the description is being cut, not by how much. Regression tests added for all three modes plus the imitator case.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 3]
```diff
@@ -710,6 +705,21 @@ export async function injectSkillCatalog(
...
     if (catalog) {
+      /**
+       * A description the catalog kept intact appears in it verbatim, so
+       * containment detects truncation from any rung of the ladder without
+       * needing entry boundaries. Boundaries are not recoverable anyway:
+       * descriptions may contain newlines, so an entry is not one physical
+       * line, and a continuation line can imitate the next entry's marker.
+       */
+      for (const s of catalogVisibleSkills) {
+        if (catalog.includes(s.description)) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Confirmed and fixed in 0d8b726. Reproduced your exact example — a skill named `research` plus a skill described `research`, forced to names-only: bare containment held and the warning was skipped even though every description was dropped.

Fixed by matching the whole `- name: description` rendering rather than the description alone. Still containment, so no entry parsing and the multiline / duplicate-name / imitator cases from the earlier rounds stay correct — but a collision now has to reproduce the entry's full rendering, not just its text.

Checked all five adversarial cases against the real formatter. Entry-scoped containment is correct on 5/5; the positional version was wrong on 3, bare containment wrong on 1 (yours). Regression test added for the collision.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 4]
```diff
@@ -583,6 +585,25 @@ export async function resolveSkillCatalog(
...
+ * another skill's name once the catalog falls back to names-only.
+ *
+ * Matching the rendering rather than parsing entries also keeps this correct
+ * where boundaries are not recoverable: descriptions may contain newlines, so
+ * an entry is not one physical line, and a continuation line can imitate the
+ * next entry's marker.
+ */
+function catalogKeptDescription(catalog: string, name: string, description: string): boolean {
+  const entry = `\n- ${name}: ${description}`;
+  return catalog.includes(`${entry}\n`) || catalog.endsWith(entry);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Confirmed and fixed in 31ddf33, and you were right to push on identity rather than accept another patch.

Reproduced it: a description of 249 chars + U+2026 + `\n- next-skill: next description`, with a real `next-skill` entry matching, leaves the full rendering present across two entries and suppresses the warning.

That was the fourth variant of one problem, so I stopped patching the check and moved the measurement somewhere unambiguous. Every decision in the truncation ladder reads description `.length` and never content (`skillCatalog.ts` lines 51-89), so formatting a probe whose descriptions are same-length filler reproduces the real cuts exactly — and the probe is parseable by construction, since filler carries no newline or entry marker and skill names are validated to `^[a-z0-9][a-z0-9-]*$`.

No formatter metadata needed, and it closes the class rather than the instance: verified against all six cases raised across these four rounds, correct on 6/6. It also restores the delivered length to the warning, which the containment approach had cost.

**Comment by danny-avila** [comment]
@codex review
